accounts-db/src/accounts.rs

```diff
@@ -44,12 +44,37 @@ use {
     },
 };
 
+// max number of accts an entry can hold: 64 * 256
+const ENTRY_ACCTS_LOOKUP_TABLE_SIZE: usize = 16384;
 pub type PubkeyAccountSlot = (Pubkey, AccountSharedData, Slot);
 
+#[derive(Debug, Default)]
+struct BatchAccountLocks {
+    writables: HashSet<Pubkey>,
+    readables: HashSet<Pubkey>,
+}
+
+impl BatchAccountLocks {
+    fn with_capacity(capacity: usize) -> Self {
+        BatchAccountLocks {
+            writables: HashSet::with_capacity(capacity),
+            readables: HashSet::with_capacity(capacity),
+        }
+    }
+
+    fn insert_read_lock(&mut self, key: &Pubkey) {
+        self.readables.insert(*key);
+    }
+
+    fn insert_write_lock(&mut self, key: &Pubkey) {
+        self.writables.insert(*key);
+    }
+}
+
 #[cfg_attr(feature = "frozen-abi", derive(AbiExample))]
 #[derive(Debug, Default)]
 pub struct AccountLocks {
-    write_locks: HashSet<Pubkey>,
+    write_locks: HashMap<Pubkey, u64>,
```

Comment:
we should have a code comment on why it's possible to have "multiple" outstanding write-locks.